Adding random fact domain layer

SamilaRuane opened the pull request:
### :pushpin: Referências

* **Repositório da Atividade:** [Ex03](https://github.com/stone-payments/mobile-training/blob/master/ex03-chucknorris.md) 

### :cyclone: Mudanças

Este PR adiciona a camada de domínio da feature que contempla a recuperação de um fato randômico do serviço e define contratos de comunicação com esta camada.
 
### :art: Mudanças de UI
_Não foram adicionadas mudanças de ui_

### :memo: Notas

NA

SamilaRuane committed:
Adding use case class and feature contracts.

AllanHasegawa commented:
Esse PR ainda está com os commits da branch de layout, tem como arrumar ele para mostrar só os commits novos?

Talvez vc precise dar um squash e dai um rebase com a feature branch.

AllanHasegawa commented on domain/src/test/java/br/stone/mobiletraining/samilasantos/domain/randomFact/RandomFactMother.kt:
```diff
@@ -0,0 +1,52 @@
+    val repository: RandomFactRepository = mock(RandomFactRepository::class.java)
+
+    fun withASuccessScenario(func: () -> Unit) {
+        `when`(repository.getFact())
```
O Mockito do Kotiln tem essa função como `whenever`, que fica mais fácil de digitar do que esses "ticks".

AllanHasegawa commented on domain/src/main/java/br/stone/mobiletraining/samilasantos/domain/randomFact/RandomFactRepository.kt:
```diff
@@ -0,0 +1,8 @@
+import io.reactivex.Observable
+
+interface RandomFactRepository {
+    fun getFact(): Observable<Fact>
```
Falta uma documentação de como essa stream vai funcionar. Caso ela emite apenas um item e já termina, seria bom ela retornar um `Single<Fact>`, que dai fica claro para quem tá chamando o comportamento dela.

AllanHasegawa commented on domain/src/main/java/br/stone/mobiletraining/samilasantos/domain/common/IntegrationErrors.kt:
```diff
@@ -0,0 +1,7 @@
+package br.stone.mobiletraining.samilasantos.domain.common
+
+sealed class IntegrationErrors : Throwable() {
+    object FactNotFound : IntegrationErrors()
```
Eu não colocaria `FactNotFound` como um "common" na camada de domain. Acho que ficaria melhor deixar essa exception em uma feature específica.

Falando em exceptions. Acho que chamar esses carinha de "Errors" pode não ser correto. Veja esse link: https://javaconceptoftheday.com/difference-between-error-vs-exception-in-java/

Penso que "Exceptions" aqui seria mais apropriado.

AllanHasegawa commented on domain/src/test/java/br/stone/mobiletraining/samilasantos/domain/randomFact/RandomFactTest.kt:
```diff
@@ -0,0 +1,75 @@
+class RandomFactTest {
+
+    @Test
+    fun shouldReturnANoNetworkError() {
```
Em testes unitários podemos deixar os métodos mais claros com uma String no padrão "given X then Y should be Z" ou algo assim xD

Por exemplo:

```
@Test
fun `given a no network error, GetRandomFact should return an error`() {
...
}
```

SamilaRuane committed:
Adding some adjustments

Changing Mockito dependency to kotlin mockito.
Using whenever instead when.
Changing uc return to Single instead Observable
Separating exceptions.

AllanHasegawa reviewed (approved):
:rocket:
